### PR TITLE
1465 review geocoding stats

### PR DESCRIPTION
**Pull request by juanignaciosl**
@rafatower CR this fixing #1465, please. I haven't tested it in staging yet (can't test locally), but I first want to know whether this is what you're expecting.

### CODE REVIEW

**Comment by rafatower** [comment]
I'm afraid that will not solve the problem: geocodings triggered from the georeferencer won't go through the model you modified. See [these lines](https://github.com/CartoDB/cartodb/blob/master/services/importer/lib/importer/georeferencer.rb#L162-L206)

**Comment by juanignaciosl** [comment]
I'm not sure about what you mean. AFAIK there're two ways you can trigger geocodings: UI and georeferencer. The new flag is set when user requests geocoding manually, and other cases won't set the flag. That way we can distinguish them. Or, at least, that's what I was trying!

**Comment by rafatower** [comment]
But we won't get logs from the georeferencer in kibana, right?

**Comment by juanignaciosl** [comment]
Oh, I think I see what you mean :confused:  I had read the lines you pointed, but I supposed [CartoDB::InternalGeocoder::Geocoder.new](https://github.com/CartoDB/cartodb/blob/master/services/importer/lib/importer/georeferencer.rb#L179) creates a new Geocoding instance, but it doesn't, does it? Internal geocoding is neither triggering metrics not registering the execution? :disappointed:

**Comment by rafatower** [comment]
:(

As far as I remember it is the other way around: the Geocoding creates a CartoDB::InternalGeocoder::Geocoder (see [table_geocoder](https://github.com/CartoDB/cartodb/blob/ec4d3ea68680774be6cd4de2fa4d02d8d0d4454d/app/models/geocoding.rb#L53)). And it is the Geocoding also responsible of the so-called [metrics_payload](https://github.com/CartoDB/cartodb/blob/ec4d3ea68680774be6cd4de2fa4d02d8d0d4454d/app/models/geocoding.rb#L184)

File: app/models/geocoding.rb [hunk 1]
```diff
@@ -93,6 +108,7 @@ def run!
       puts "Processed: #{processed_rows}"
       raise 'Geocoding timeout' if Time.now - started > run_timeout and ['started', 'submitted', 'accepted'].include? state
       raise 'Geocoding failed'  if state == 'failed'
+      # TODO: what is this sleep for?
```
Comment (juanignaciosl) [inline app/models/geocoding.rb]: @rafatower do you know what is this sleep for? :confused:

**Comment by juanignaciosl** [comment]
@rafatower CR this, please. Now it should be much closer to an actual solution :-) I haven't tested it in staging yet, I want to know your opinion before.

File: app/models/geocoding.rb (hunk 1 above)
Comment (rafatower) [inline app/models/geocoding.rb]: I think this avoids writing state information in the db model too fast

File: services/importer/lib/importer/georeferencer.rb [hunk 2]
```diff
@@ -160,9 +160,18 @@ def create_the_geom_from_ip_guessing
...
+      def geocode_ips ip_column_name
+        job.log "Geocoding ips..."
+        geocode(ip_column_name, 'point', 'ipaddress')
+      end
+
+      def geocode(formatter, geometry_type, kind)
         geocoder = nil
-        @importer_stats.timing('geocoding') do
-          job.log "Geocoding countries..."
+        @importer_stats.timing("geocoding.#{kind}") do
```
Comment (rafatower) [inline services/importer/lib/importer/georeferencer.rb]: out of scope but still: I think we're producing a lot of perf data that is rather hard to exploit

File: services/importer/lib/importer/georeferencer.rb [hunk 3]
```diff
@@ -170,41 +179,48 @@ def geocode_countries country_column_name
...
             max_rows: nil,
             country_column: nil
           )
           geocoder = CartoDB::InternalGeocoder::Geocoder.new(config)
-          geocoder.run
-          @tracker.call('importing')
+
+          begin
+            geocoding = Geocoding.new config.slice(:kind, :geometry_type, :formatter, :table_name)
+            geocoding.force_geocoder(geocoder)
```
Comment (rafatower) [inline services/importer/lib/importer/georeferencer.rb]: Not 100% sure but maybe the Geocoding class is able to create the geocoder on its own, therefore avoiding the need of creating and forcing it explicitly. What do you think?

File: spec/models/geocoding_spec.rb [hunk 4]
```diff
@@ -96,7 +96,7 @@
       geocoding.table_geocoder.stubs(:run).returns true
       geocoding.table_geocoder.stubs(:cache).returns  OpenStruct.new(hits: 5000)
       geocoding.table_geocoder.stubs(:process_results).returns true
-      geocoding.stubs(:processable_rows).returns 10
+      geocoding.class.stubs(:processable_rows).returns 10
```
Comment (rafatower) [inline spec/models/geocoding_spec.rb]: nice, thanks!

File: services/importer/lib/importer/georeferencer.rb (hunk 3 above)
Comment (rafatower) [inline services/importer/lib/importer/georeferencer.rb]: No, I'm wrong. At this point the table is not fully imported and, for instance, has a temp name

**Comment by rafatower** [comment]
:+1:

**Pull request merged**